### PR TITLE
Fix DeterminePartitionsJob reducer when total rows < targetPartitionSize * SHARD_COMBINE_THRESHOLD

### PR DESCRIPTION
**Pull request by jon-wei**
This PR fixes an `ArrayIndexOutOfBoundsException` when a `targetPartitionSize` is specified but a DeterminePartitionsJob reducer receives an iterable where the total # rows is less than `targetPartitionSize * SHARD_COMBINE_THRESHOLD` (0.25).

```
2019-08-08T02:08:11,206 WARN [main] org.apache.hadoop.mapred.YarnChild - Exception running child : java.lang.ArrayIndexOutOfBoundsException: -1
	at java.util.ArrayList.elementData(ArrayList.java:422)
	at java.util.ArrayList.remove(ArrayList.java:499)
	at org.apache.druid.indexer.DeterminePartitionsJob$DeterminePartitionsDimSelectionReducer.innerReduce(DeterminePartitionsJob.java:712)
	at org.apache.druid.indexer.DeterminePartitionsJob$DeterminePartitionsDimSelectionBaseReducer.reduce(DeterminePartitionsJob.java:562)
	at org.apache.druid.indexer.DeterminePartitionsJob$DeterminePartitionsDimSelectionBaseReducer.reduce(DeterminePartitionsJob.java:538)
	at org.apache.hadoop.mapreduce.Reducer.run(Reducer.java:171)
	at org.apache.hadoop.mapred.ReduceTask.runNewReducer(ReduceTask.java:627)
	at org.apache.hadoop.mapred.ReduceTask.run(ReduceTask.java:389)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1836)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)
```

### CODE REVIEW

File: indexing-hadoop/src/main/java/org/apache/druid/indexer/DeterminePartitionsJob.java [hunk 1]
```diff
@@ -691,8 +691,9 @@ protected void innerReduce(Context context, SortableBytes keyBytes, Iterable<Dim
             // One more shard to go
             final ShardSpec shardSpec;
 
-            if (currentDimPartition.rows < config.getTargetPartitionSize() * SHARD_COMBINE_THRESHOLD) {
-              // Combine with previous shard
+            if (currentDimPartition.rows < config.getTargetPartitionSize() * SHARD_COMBINE_THRESHOLD &&
+                currentDimPartitions.partitions.size() > 0) {
```
Comment (ccaominh) [inline indexing-hadoop/src/main/java/org/apache/druid/indexer/DeterminePartitionsJob.java]: Consider `!isEmpty()` over `size() > 0`: https://rules.sonarsource.com/java/tag/clumsy/RSPEC-1155

Comment (jon-wei) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/DeterminePartitionsJob.java]: Changed to use !isEmpty()

Comment (jon-wei) [reply indexing-hadoop/src/main/java/org/apache/druid/indexer/DeterminePartitionsJob.java]: weird, GH didn't pick up my latest commit

**Comment by gianm** [comment]
I restarted the "other integration test" job, which had blown up.

**Pull request merged**